src/wx/widgets/sdljoy.cpp

```diff
@@ -38,420 +39,489 @@ static int16_t axisval(int16_t x)
+            // Do nothing. This will be handled with JOYDEVICEADDED and
+            // JOYDEVICEREMOVED events.
+            break;
+
+        // Joystick events for non-GameControllers.
+        case SDL_JOYBUTTONDOWN:
+        case SDL_JOYBUTTONUP:
+        {
+            SDL_JoystickID joy_id = e.jbutton.which;
+            auto joy_state = joystick_states_.find(joy_id);
+            if (joy_state != joystick_states_.end()) {
+                // Do not process joystick events for game controllers.
+                if (joy_state->second->IsGameController())
+                    break;
 
-                break;
+                joy_state->second->ProcessEvent(
+                    e.type, e.jbutton.button, e.jbutton.state);
             }
-            case SDL_JOYDEVICEREMOVED:
-            {
-                auto joy = e.jdevice.which;
 
-                if (contains(instance_map, joy)) {
-                    auto index = instance_map[joy]->index;
-                    RemapControllers();
-
-                    systemScreenMessage(wxString::Format(_("Disconnected joystick %d"), index));
-                }
-
-                got_event = true;
-
-                break;
-            }
+            got_event = true;
+            break;
         }
-    }
 
-    bool do_poll  = false;
-    wxLongLong tm = wxGetUTCTimeMillis();
-
-    if (got_event)
-        last_poll = tm;
-    else if (tm - last_poll > POLL_TIME_MS) {
-        do_poll   = true;
-        last_poll = tm;
-    }
-
-    if (do_poll) {
-        for (auto&& joy : joystate) {
-            if (!joy.second.dev) continue;
-
-            if (SDL_IsGameController(joy.first)) {
-                for (uint8_t but = 0; but < SDL_CONTROLLER_BUTTON_MAX; but++) {
-                    auto last_state = joy.second.button[but];
-                    auto state      = SDL_GameControllerGetButton(joy.second.dev, static_cast<SDL_GameControllerButton>(but));
-
-                    if (last_state != state) {
-                        CreateAndSendEvent(joy.first, WXSDLJOY_BUTTON, but, state, last_state);
-
-                        joy.second.button[but] = state;
-
-                        wxLogDebug("POLLED SDL_CONTROLLERBUTTON: joy:%d but:%d val:%d prev_val:%d", joy.first, but, state, last_state);
-                    }
-                }
-
-                for (uint8_t axis = 0; axis < SDL_CONTROLLER_AXIS_MAX; axis++) {
-                    auto val      = axisval(SDL_GameControllerGetAxis(joy.second.dev, static_cast<SDL_GameControllerAxis>(axis)));
-                    auto prev_val = joy.second.axis[axis];
-
-                    if (val != prev_val) {
-                        CreateAndSendEvent(joy.first, WXSDLJOY_AXIS, axis, val, prev_val);
-
-                        joy.second.axis[axis] = val;
+        case SDL_JOYAXISMOTION:
+        {
+            SDL_JoystickID joy_id = e.jaxis.which;
+            auto joy_state = joystick_states_.find(joy_id);
+            if (joy_state != joystick_states_.end()) {
```

Comment:
I was trying to do that but my local checkout had issues finding `<optional>` I am using mingw to build and test on Windows, it might have something to do with that. In any case, feel free to do it as I have no way of testing my code is correct.

Reply:
Which version of mingw are you using in what environment? Which version of gcc is it?